### ISSUE TITLE
Show subfolders as real subfolders in folder list

### ISSUE THREAD

**Issue by Ede123** [thunderbird/thunderbird-android#630]
Currently all folders are shown on the same level with names like

```
Folder1
Folder1.Subfolder1
Folder1.Subfolder2
```

Actually it would be much clearer if the folder tree wouldn't be flat, e.g. subfolders where on a deeper level and reachable by tapping the containing parent folder.

Especially with many folders the current folder lists becomes almost unusable.

**Comment by dewi-ny-je** [comment]
I just connected K9 Mail to my IMAP server at home, where I stock my archived mail, and the issue mentioned is much felt. Please know that I also vote in favour of the enhancement.

**Comment by RPDiep** [comment]
I would also vote for this. It's on the wish list for quite some time:
http://code.google.com/p/k9mail/issues/detail?id=17

**Comment by philipwhiuk** [comment]
If someone knows of a good library that allows:

* Visiting tree node entries
* Expanded and collapsing leaves

then this will be fairly straightforward to do - just replacing the existing list component

If not we will probably have to roll our own which is time consuming to develop.

I looked at [Android Tree View](https://android-arsenal.com/details/1/1534) for example but it doesn't seem to allow you to visit nodes which aren't leaves (e.g. there'd be no way to visit Folder1).

One other (additional?) alternative is to have a view where you have the sub folders of a folder displayed in a section above the email list as in the following ASCII art

```
-----               Folder 1             -------
-------------------------------------------------
------------------ Subfolder1 ------------------ 
------------------ Subfolder2 ------------------ 
------------------ Subfolder3 ------------------ 
------------------------------------------------
First email subject ----- Smith ----- 13/1/2009
Second email subject ---- Smith ----- 14/1/2009
```

**Comment by john-walsh** [comment]
tip:

Do use the standard 'Android Tree View', but you need to create 'self' leaves.

eg:

node: Folder1
leaf: Folder1.self
node: Folder1.Subfolder1
leaf: Folder1.Subfolder1.self
node: Folder1.Subfolder2
leaf: Folder1.Subfolder2.self

**Comment by Valodim** [comment]
Folders in IMAP that have subfolders can't also contain email, so android tree view should be fine. This is actually relatively simple then:

- translate folder delimiters
- replace folder list widget with tree view

Since this is just a matter of how things are displayed, I don't think we need actual database structure for subfolders to do this. If someone wants to give it a go, these can be tackled in any order :)

**Comment by dewi-ny-je** [comment]
I have IMAP folders containing also both and further subfolders. I use dovecot as server.

**Comment by cketti** [comment]
@dewi-ny-je is right. There can be folders that can't contain messages, but it's not a requirement for folders that contain subfolders.

I'd like us to store a proper tree structure in the database. In addition to that I'd like us to store the original (utf7-encoded) folder name in the database. That way we don't have to fetch the delimiter character and encode folder names all the time. It will also allow us to support folder names which aren't valid utf7 sequences.

**Comment by sebalis** [comment]
I am a beginning Android user, so please forgive me if I get terminology wrong or miss something obvious. Also, I’m using the “Material design” variant of K-9; but I believe what I’m going to describe applies in the default version too.

In my view a special tree view is not required or even beneficial for this feature (which I too very much want to be added). The existing list view has something that I will call sublists or nested lists: tapping an account brings up its folders, tapping a folder brings up its contents, and the Back button always returns to the previous view. It would be very logical if subfolders were shown using sublists too. The only question is how to show a mail folder that has both emails in it as well as subfolders below it.

There are several answers to that question, I have come up with four and I would suggest providing account options to choose between them: 1. show emails on the same level as subfolders, emails on top, 2. same as 1. but subfolders on top, 3. show subfolders plus a pseudo-subfolder on top named “(_n_&nbsp;emails)” – shown only if _n_>0 – so another tap is required to actually get to those emails; 4. the other way round, i.e. show emails plus a pseudo-email on top named “(_n_&nbsp;subfolders)” – again shown only if _n_>0 – so another tap is required to navigate into subfolders.

I think that implementing it in this way would be cleaner as it uses an interface that users are familiar with. Also, in this way subfolders would not need to be indented, so the full screen width can be used to display their names.

**Comment by sebalis** [comment]
I’m now wondering if what I am talking about is the “standard Android Tree View” that @john-walsh mentioned [above](#issuecomment-269362006), but hopefully I was still able to add something …

**Comment by dewi-ny-je** [comment]
Yes the solution suggested by @sebalis should require the least amount of work and would solve completely the issue.

**Comment by john-walsh** [comment]
Thanks @sebalis  and  @dewi-ny-je - yes, the solution is the same underlying mechanisum, and there are a couple of ways to deal with 'leaves' not being able to hold data (emails), so this should not be hard to implement.

**Comment by IanBod** [comment]
I too would very much welcome some form of tree structure for the folder view. The exact form doesn't matter provided subfolders can be hidden and displayed again as required. 

The solution offered by @sebalis seems sensible. It could include the message count after the (sub)folder name which deals with the issue of folders containing both emails and subfolders.

**Comment by jroldroyd** [comment]
I have almost 900 folders on my IMAP server, so this would be very much appreciated.  Not just in the Folders list, but also in the Choose Folder dialog when moving messages, etc.

**Comment by zerpex** [comment]
+1 this is very annoying and the only things that is very much missing for me.

**Comment by sebma** [comment]
@zerpex   I agree, this is the only feature I really miss
+1

**Comment by kikig1** [comment]
Are there any news since 2015? A Treeview for the Folders would really be very helpful ;)

**Comment by finex** [comment]
Hi, this feature request is active since 2008. Probably it will never be implemented.

**Comment by nomis** [comment]
I think the folders would need to be displayed with two touch areas - one that opens the folder and another that either:
* Expands/collapses it in place
* Navigates into that sub-section of the folder tree, displaying only those sub-folders until the back button is used

**Comment by ASLLR** [comment]
Hi everyone, 

we have an idea when this functionnality is implemented ? why is this so long ?

Thank you

**Comment by ecxod** [comment]
> If someone knows of a good library that allows:
> 
>     * Visiting tree node entries
> 
>     * Expanded and collapsing leaves


There are several Android libraries that allow visiting tree node entries, expanding and collapsing. Some popular ones include: 

- [Android TreeView](https://github.com/bmelnychuk/AndroidTreeView): This library provides a tree view widget for Android. It supports expanding and collapsing nodes, as well as customizing the appearance of the tree view.

- [TreeViewList](https://github.com/bongni/TreeViewList): This library provides a simple tree view for Android, with support for expanding and collapsing nodes, and might be a good fit for displaying hierarchical data such as an IMAP folder set.

- [ExpandableRecyclerView](https://github.com/thoughtbot/expandable-recycler-view): This library provides an expandable version of Android's RecyclerView widget, with support for expanding and collapsing nodes.

- [TreeView](https://github.com/TellH/RecyclerTreeView): This library provides a flexible and customizable tree view for Android, with support for expanding and collapsing nodes.

Honestly, I was hoping it would be the first thing thunderbird does. No other email app shows me my list of 23487 IMAP folders as one list.

**Comment by CekLive** [comment]
> Attualmente tutte le cartelle sono visualizzate sullo stesso livello con nomi come
> 
> ```
> Folder1
> Folder1.Subfolder1
> Folder1.Subfolder2
> ```
> 
> In realtà sarebbe molto più chiaro se l'albero delle cartelle non fosse piatto, ad esempio le sottocartelle fossero a un livello più profondo e raggiungibili toccando la cartella principale che le contiene.
> 
> Soprattutto con molte cartelle, gli elenchi delle cartelle correnti diventano quasi inutilizzabili.


I agree, K-9 will be used by Thunderbird users, so reverting some of the original Thunderbird interface would be logical
![Tree](https://user-images.githubusercontent.com/127445378/224382718-11300737-45e5-4f04-a920-e8094b10a8db.jpg)

**Comment by Dave-Rado** [comment]
According to the [Roadmap](https://developer.thunderbird.net/planning/android-roadmap), (which confusing calls itself the Thunderbird roadmap but it is really the K-9 roadmap), this should have been implemented in Q1 2023 - but it is already well into Q2, a new version of K-9 has just been released and this still hasn't been implemented. What's happening with this?

Also, it's not just making the folder structure collapsed by default and expandable that's required in order to make the folders easily navigable. It's also making the folder names display in full with wordwrapping, and taking up the full width of the screen - as already happens when you select "Manage Folders". Without this, it will still be a pain to navigate even after you've made the folders collapsed and expandable.

**Comment by HoundDogRock** [comment]
Still not implemented as of february 2024. 
I have a LOT of folders and sub-folders, so the application is not usable without expandable sub-folders.

**Comment by Gerben321** [comment]
It's impressive this still isn't five after 9 years. I doubt that road map is updated because there are more things on there that still haven't happened.

**Comment by ecxod** [comment]
It is related to understanding maildir format. The developer of K-9 does not understand the maildir "folder" format.

**Comment by ecxod** [comment]
> Currently all folders are shown on the same level with names like
> 
> ```
> Folder1
> Folder1.Subfolder1
> Folder1.Subfolder2
> ```
> 
> Actually it would be much clearer if the folder tree wouldn't be flat, e.g. subfolders where on a deeper level and reachable by tapping the containing parent folder.
> 
> Especially with many folders the current folder lists becomes almost unusable.

The main Problem is that the developer does not understand the IMAP maildir folder structure

.Folder1
.Folder1.Subfolder1
.Folder1.Subfolder2

(please pay attention to the leading dot)

IMAP Servers who are set to `maildir` format, display folders delimited by `.`, in the above shown way. `.Folder1.Subfolder1.SubSubFolder1`. Where `.XXX` is a folder element.
Please do not confuse the setting in `10-mail.conf` called `maildir_separator` that is set to `/` with this dot. This `/` is just to tell the client to display the folder in a nested way.

The developer of K-9 has a different concept about this. In his world there are important folders (that contain emails about walking dog topics) and not important folders (all the rest).

have fun.

**Comment by bjo81** [comment]
@ecxod Sorry, but this is a stupid comment. Also with dovecot's mdbox it's `.Foo.Bar`. Other Android Clients like e.g. FairMail are able to use some tree view.

**Comment by RamonAbudAlcala** [comment]
I'd like to vote for this too! Please implement!

**Comment by MarcusWolschon** [comment]
My old folder list, that at least had a SEARCH function to have any chance to find one of the 200+ folders on my IMAP server,
was replaced with the sidebar ages ago.
I have to go to a Desktop to get any mail outside the 10 most important folders.
"move to folder" still has a search feature but the folder list hasn't.

A tree view is absolutely NEEDED.

The cited roadmap only contains "ability to manage and organize long folder lists." now, wich only speaks about managing folders not READING EMAILS in them.

**Comment by kewisch** [comment]
Ultimately fixed by #9045, but based on the great work started in #8956

**Comment by ClemRiviere** [comment]
It's been a pleasure to have been able to move the subject forward and work with you :)

**Comment by Dave-Rado** [comment]
When will this go live? My copy of k-9 mail hasn't changed.

**Comment by kewisch** [comment]
It is currently behind a feature flag as we're making a few more changes to the drawer, but if all goes well it will be in Thunderbird 11. We're currently checking to see if we can uplift the changes to the next beta, so if you want to see it early make sure you have https://play.google.com/store/apps/details?id=net.thunderbird.android.beta installed :)

**Comment by john-walsh** [comment]
Thank you @ClemRiviere , and all that helped out.

I look forward to version 11.
